UIforETW/PowerStatus.cpp

```diff
@@ -108,125 +108,179 @@ void CPowerStatusMonitor::SampleBatteryStat()
-										}
-
-										float batPercentage = bs.Capacity * 100.f / bi.FullChargedCapacity;
-
-										char rate[100];
-										if (bs.Rate == BATTERY_UNKNOWN_RATE)
-											sprintf_s(rate, "Unknown rate");
-										else if (bi.Capabilities & BATTERY_CAPACITY_RELATIVE)
-											sprintf_s(rate, "%ld (unknown units)", bs.Rate);
-										else
-											sprintf_s(rate, "%1.3f watts", bs.Rate / 1000.0f);
-										ETWMarkBatteryStatus(powerState, batPercentage, rate);
-										//outputPrintf(L"%S, %1.3f%%, %S\n", powerState, batPercentage, rate);
-									}
-								}
-							}
-						}
-						CloseHandle(hBattery);
-					}
-				}
-				LocalFree(pdidd);
-			}
+												NULL);
+
+		if (!(tagQueried && bqi.BatteryTag))
+		{
+			FreeLocalAlloc(pdidd);
+			CloseValidHandle(hBattery);
+			continue;
+		}
+
+		// With the tag, you can query the battery info.
+		BATTERY_INFORMATION bi = {};
+		bqi.InformationLevel = BatteryInformation;
+		
+		const BOOL infoQueried = DeviceIoControl(hBattery,
+												 IOCTL_BATTERY_QUERY_INFORMATION,
+												 &bqi,
+												 sizeof(bqi),
+												 &bi,
+												 sizeof(bi),
+												 &dwOut,
+												 NULL);
+		if (!infoQueried)
+		{
+			FreeLocalAlloc(pdidd);
+			CloseValidHandle(hBattery);
+			continue;
 		}
-		else  if (ERROR_NO_MORE_ITEMS == GetLastError())
+
+		// Only non-UPS system batteries count
+		if (!(bi.Capabilities & BATTERY_SYSTEM_BATTERY))
+		{
+			FreeLocalAlloc(pdidd);
+			CloseValidHandle(hBattery);
+			continue;
+		}
+
+
+		// Query the battery status.
+		BATTERY_WAIT_STATUS bws = {};
+		bws.BatteryTag = bqi.BatteryTag;
+
+		BATTERY_STATUS bs;
+		
+		const BOOL statusQueried = DeviceIoControl(hBattery,
+												   IOCTL_BATTERY_QUERY_STATUS,
+												   &bws,
+												   sizeof(bws),
+												   &bs,
+												   sizeof(bs),
+												   &dwOut,
+												   NULL);
+		if (!statusQueried)
 		{
-			break;  // Enumeration failed - perhaps we're out of items
+			FreeLocalAlloc(pdidd);
+			CloseValidHandle(hBattery);
+			continue;
 		}
+
+		char powerState[100];
+		powerState[0] = 0;
+		if (bs.PowerState & BATTERY_CHARGING)
+			strcat_s(powerState, "Charging");
+		if (bs.PowerState & BATTERY_DISCHARGING)
+			strcat_s(powerState, "Discharging");
+		if (bs.PowerState & BATTERY_POWER_ON_LINE)
+		{
+			if (powerState[0])
+				strcat_s(powerState, ", on AC power");
+			else
+				strcat_s(powerState, "On AC power");
+		}
+
+		float batPercentage = bs.Capacity * 100.f / bi.FullChargedCapacity;
+
+		char rate[100];
+		if (bs.Rate == BATTERY_UNKNOWN_RATE)
+			sprintf_s(rate, "Unknown rate");
+		else if (bi.Capabilities & BATTERY_CAPACITY_RELATIVE)
+			sprintf_s(rate, "%ld (unknown units)", bs.Rate);
+		else
+			sprintf_s(rate, "%1.3f watts", bs.Rate / 1000.0f);
+		ETWMarkBatteryStatus(powerState, batPercentage, rate);
+		//outputPrintf(L"%S, %1.3f%%, %S\n", powerState, batPercentage, rate);
+		CloseValidHandle(hBattery);

```

Comment:
Way too many calls to CloseValidHandle and FreeLocalAlloc(pdidd). That much copy/paste means that it is definitely for resource management classes. std::vector for the memory, and some handle wrapper class for the handle. That's nicer than adding wrapper functions anyway.